### PR TITLE
[Feature][DatePicker] Component

### PR DESCRIPTION
**Pull request by bumjoo**
## Description

A Date picker component
* This component provide a simple way to select a date value from calendar popup.
* You can customize the year and month area on the calendar.
* You can set the type of day display. ('narrow' | 'short')

dooboo_datepicker

## Related Issues

#305 [Feature] DatePicker

## Tests

Simple rendering snapshot tests.

## Checklist

Before you create this PR confirms that it meets all requirements listed below by checking the relevant checkboxes (`[x]`). This will ensure a smooth and quick review process.

- [x] I read the [Contributor Guide](https://github.com/dooboolab/dooboo-ui/blob/master/CONTRIBUTING.md) and followed the process outlined there for submitting PRs.
- [x] Run `yarn test` or `yarn test -u` if you need to update snapshot.
- [x] Run `yarn lint`
- [x] I am willing to follow-up on review comments in a timely manner.

### CODE REVIEW

File: stories/dooboo-ui/DatePicker.stories.tsx [hunk 1]
```diff
@@ -0,0 +1,141 @@
+import { IC_FACEBOOK, IC_GOOGLE } from '../Icon';
```
Comment (JeffGuKang) [inline stories/dooboo-ui/DatePicker.stories.tsx]: Delete the line.

File: stories/dooboo-ui/DatePicker.stories.tsx [hunk 2]
```diff
@@ -0,0 +1,141 @@
...
+  justify-content: center;
+  width: 100%;
+  margin-top: 20px;
+  margin-bottom: 20px;
+  flex-direction: column;
+`;
+
+function Default(): React.ReactElement {
+  const [facebookLoading, setFacebookLoading] = useState<boolean>(false);
+  const [googleLoading, setGoogleLoading] = useState<boolean>(false);
```
Comment (JeffGuKang) [inline stories/dooboo-ui/DatePicker.stories.tsx]: It looks useless states.

File: main/DatePicker/README.md [hunk 3]
```diff
@@ -0,0 +1,32 @@
+# DatePicker
+
+> [DatePicker] Simple datepicker component for react-native. This componenet supports select the date on the calendar.
+
+## Preview
+![dooboo_datepicker](https://user-images.githubusercontent.com/38123162/92732197-11b42100-f3b1-11ea-9c11-ac9dd7aec5e5.gif)
+## Props
```
Comment (JeffGuKang) [inline main/DatePicker/README.md:9]: There are no descrpitions about `yearMonthComponent`, `weekdayFormat` and so on.

File: main/DatePicker/README.md [hunk 4]
```diff
@@ -0,0 +1,32 @@
...
+```sh
+yarn add dooboo-ui
+```
+
+## Getting started
+- Import
+
+```typescript
+import { Datepicker } from 'dooboo-ui';
+```
```
Comment (JeffGuKang) [inline main/DatePicker/README.md:66]: Keep the clean MARKDOWN rule as https://raw.githubusercontent.com/dooboolab/dooboo-ui/master/main/PinchZoom/README.md.

File: stories/dooboo-ui/DatePicker.stories.tsx [hunk 5]
```diff
@@ -0,0 +1,141 @@
+import { IC_FACEBOOK, IC_GOOGLE } from '../Icon';
+import { Image, Text, View } from 'react-native';
+import React, { ReactElement, useState } from 'react';
+
+import { ContainerDeco } from '../../storybook/decorators';
+import { DatePicker } from '../../main';
+import { action } from '@storybook/addon-actions';
```
Comment (JeffGuKang) [inline stories/dooboo-ui/DatePicker.stories.tsx]: Useless import.

File: stories/dooboo-ui/DatePicker.stories.tsx [hunk 6]
```diff
@@ -0,0 +1,141 @@
...
+import { storiesOf } from '@storybook/react-native';
+import styled from 'styled-components/native';
+import { text } from '@storybook/addon-knobs';
```
Comment (JeffGuKang) [inline stories/dooboo-ui/DatePicker.stories.tsx]: Useless import.

File: main/DatePicker/README.md [hunk 7]
```diff
@@ -0,0 +1,32 @@
...
+|selectedDate          |                |Date               |`new Date()`
+| dateInputStyle           |           | ViewStyle                | `true`                |
+| style        |           | ViewStyle                 |                   |
+| label        |           | string     | `default arrow image` |
+| labelTextStyle   |           | TextStyle              | ``                    |
+| errorText           |           | string              | `Invalid Date`                    |
+| errorTextStyle            |           | TextStyle              | `{ color: '#F00', textAlign: 'left' }`                    |
+|dateTextStyle        |                  | TextStyle            |
+|placeholder        |       |string|`YYYY-MM-DD`|
+|placeholderTextColor|      |string|            |
```
Comment (JeffGuKang) [inline main/DatePicker/README.md]: Please fix their indentation.

File: main/DatePicker/index.tsx [hunk 8]
```diff
@@ -0,0 +1,71 @@
...
+  const [selectedDate, setSelectedDate] = React.useState<Date>(
+    props.selectedDate || new Date(),
+  );
+  const [calendarVisible, setCalendarVisible] = React.useState<boolean>(false);
+
+  const onSelectDate = (newDate: Date): void => {
+    setSelectedDate(newDate);
+    setCalendarVisible(false);
+    props.onChangeDate && props.onChangeDate(newDate);
+  };
```
Comment (JeffGuKang) [inline main/DatePicker/index.tsx:38]: New line before paragraph code.

File: stories/dooboo-ui/DatePicker.stories.tsx [hunk 9]
```diff
@@ -0,0 +1,141 @@
...
+      </Container>
+      <Container>
+        <DatePicker weekdayFormat={'short'} label={'Short weekday format'} />
+      </Container>
+      <Container>
+        <DatePicker weekdayFormat={'narrow'} label={'Narrow weekday format'} />
+      </Container>
+    </ScrollContainer>
+  );
+}
```
Comment (JeffGuKang) [inline stories/dooboo-ui/DatePicker.stories.tsx:47]: New line before / after paragraph.

File: stories/dooboo-ui/DatePicker.stories.tsx [hunk 10]
```diff
@@ -0,0 +1,141 @@
...
+export const toStorybook3 = (): ReactElement => <DatePickerYearMonth />;
+toStorybook1.story = {
+  name: 'default',
+};
+toStorybook2.story = {
+  name: 'weekday format',
+};
+toStorybook3.story = {
+  name: 'custom year/month',
+};
```
Comment (JeffGuKang) [inline stories/dooboo-ui/DatePicker.stories.tsx:122]: New line before another paragraph.

File: stories/dooboo-ui/DatePicker.stories.tsx [hunk 11]
```diff
@@ -0,0 +1,141 @@
...
+                })}`}</Text>
+              </View>
+            );
+          }}
+        />
+      </Container>
+    </ScrollContainer>
+  );
+}
+/**
```
Comment (JeffGuKang) [inline stories/dooboo-ui/DatePicker.stories.tsx:103]: New line before another paragraph.

File: main/DatePicker/Calendar/Calendar.tsx [hunk 12]
```diff
@@ -0,0 +1,179 @@
...
+    const monthDate = new Date(year, month + i, 1);
+    tmpRows.push(monthDate);
+  }
+  return tmpRows;
+};
+
+interface Props<T> {
+  onChangeMonth?: (month: Date) => void; // 스와이프 하여 달 변경시 호출
+  initDate?: Date; // initial date, if undefined, it's today
+  containerStyle?: ViewStyle; // Calendar 컨테이너 스타일
```
Comment (JeffGuKang) [inline main/DatePicker/Calendar/Calendar.tsx]: Many users cannot understand Korean. Please all Korean translate to English.

File: main/DatePicker/Calendar/Calendar.tsx [hunk 13]
```diff
@@ -0,0 +1,179 @@
...
+    style,
+  }: {
+    date: Date;
+    dailyData: T; // [YYYY-MM-DD]: dailyData
+    isCurMonth: boolean;
+    isToday: boolean;
+    style?: ViewStyle;
+  }) => React.ReactElement;
+  pastRange?: number; // 과거 월 개월 수
+  futureRange?: number; // 미래 월 개월 수
```
Comment (JeffGuKang) [inline main/DatePicker/Calendar/Calendar.tsx]: Korean comments.

File: main/DatePicker/Calendar/Calendar.tsx [hunk 14]
```diff
@@ -0,0 +1,179 @@
...
+
+  calendarWidth: number; // 달 표시 width
```
Comment (JeffGuKang) [inline main/DatePicker/Calendar/Calendar.tsx]: Korean.

File: main/DatePicker/Calendar/Calendar.tsx [hunk 15]
```diff
@@ -0,0 +1,179 @@
...
+  // monthData: MonthData;
+  yearMonthComponent?: (monthFirst: Date) => React.ReactElement;
+  weekdayFormat?: 'narrow' | 'short';
+}
+/**
+ * Pure Calendar
+ * - using FlatList
+ * - infinite months
+ * - render a month view in renderItem of Flatlist
+ * - 일 컴포넌트는 dayComponent(function). props로 받음
```
Comment (JeffGuKang) [inline main/DatePicker/Calendar/Calendar.tsx]: Need to translate to English

File: main/DatePicker/Calendar/Calendar.tsx [hunk 16]
```diff
@@ -0,0 +1,179 @@
...
+ */
+
+function Calendar<T>(props: Props<T>): React.ReactElement {
+  // 요일 row height
+  // const daysRowHeight = 22;
+  // 금월 (시작월) 날짜
```
Comment (JeffGuKang) [inline main/DatePicker/Calendar/Calendar.tsx]: English please.

File: main/DatePicker/Calendar/Calendar.tsx [hunk 17]
```diff
@@ -0,0 +1,179 @@
...
+      length: props.calendarWidth,
+      offset: props.calendarWidth * index,
+      index,
+    };
+  }, []);
+
+  const MemoizedCalendarMonth = React.memo(CalendarMonth, (prev, next) => {
+    return true;
+  });
+  // 월 달력 그리기
```
Comment (JeffGuKang) [inline main/DatePicker/Calendar/Calendar.tsx]: English.

File: main/DatePicker/Calendar/Calendar.tsx [hunk 18]
```diff
@@ -0,0 +1,179 @@
...
+      if (viewableItems.length === 0) return; // 없는 경우 있음. 그냥 무시
+      const { index, isViewable, item, key } = viewableItems[0];
+      props.onChangeMonth && props.onChangeMonth(item);
+      setCurMonthFirst(item);
+    },
+    [],
+  );
+  const renderYearMonth = React.useCallback(
+    (monthFirst): React.ReactElement => {
+      console.log('>> monthFirst = ', monthFirst);
```
Comment (JeffGuKang) [inline main/DatePicker/Calendar/Calendar.tsx]: Don't leave the console log.

File: main/DatePicker/Calendar/Calendar.tsx [hunk 19]
```diff
@@ -0,0 +1,179 @@
...
+  weekdayFormat?: 'narrow' | 'short';
+}
+/**
+ * Pure Calendar
+ * - using FlatList
+ * - infinite months
+ * - render a month view in renderItem of Flatlist
+ * - 일 컴포넌트는 dayComponent(function). props로 받음
+ */
+
```
Comment (JeffGuKang) [inline main/DatePicker/Calendar/Calendar.tsx]: Delete empty line for comments.

File: main/DatePicker/Calendar/Calendar.tsx [hunk 20]
```diff
@@ -0,0 +1,179 @@
...
+        initialScrollIndex={pastRange}
+        getItemLayout={getItemLayout}
+        // onEndReachedThreshold={this.props.onEndReachedThreshold}
+        // onEndReached={this.props.onEndReached}
+        // keyboardShouldPersistTaps={this.props.keyboardShouldPersistTaps}
+      />
+      {/* {this.renderStaticHeader()} */}
+    </View>
+  );
+}
```
Comment (JeffGuKang) [inline main/DatePicker/Calendar/Calendar.tsx:156]: New line.

File: main/DatePicker/Calendar/Calendar.tsx [hunk 21]
```diff
@@ -0,0 +1,179 @@
...
+    style?: ViewStyle;
+  }) => React.ReactElement;
+  pastRange?: number; // 과거 월 개월 수
+  futureRange?: number; // 미래 월 개월 수
+
+  calendarWidth: number; // 달 표시 width
+  // calendarHeight: number;
+  // daysRowHeight: number;
+  // monthData: MonthData;
+  yearMonthComponent?: (monthFirst: Date) => React.ReactElement;
```
Comment (JeffGuKang) [inline main/DatePicker/Calendar/Calendar.tsx]: `yearMonthComponent` is a little bit wired name. 
`titleContent?: (date: Date)` or `header?: (date: Date)` will be better.

File: main/DatePicker/Calendar/Calendar.tsx [hunk 22]
```diff
@@ -0,0 +1,179 @@
...
+  weekdayFormat?: 'narrow' | 'short';
+}
```
Comment (JeffGuKang) [inline main/DatePicker/Calendar/Calendar.tsx:49]: Need a new line.

File: main/DatePicker/Calendar/Calendar.tsx [hunk 23]
```diff
@@ -0,0 +1,179 @@
...
+    return {
+      length: props.calendarWidth,
+      offset: props.calendarWidth * index,
+      index,
+    };
+  }, []);
+
+  const MemoizedCalendarMonth = React.memo(CalendarMonth, (prev, next) => {
+    return true;
+  });
```
Comment (JeffGuKang) [inline main/DatePicker/Calendar/Calendar.tsx]: const MemoizedCalendarMonth = React.memo(CalendarMonth) is same if you do not use custom comparison logic for rerender.

File: main/DatePicker/Calendar/Calendar.tsx [hunk 24]
```diff
@@ -0,0 +1,179 @@
...
+          calendarWidth={props.calendarWidth}
+          style={{
+            width: props.calendarWidth,
+          }}
+          today={initDate}
+        />
+      );
+    },
+    [],
+  );
```
Comment (JeffGuKang) [inline main/DatePicker/Calendar/Calendar.tsx:97]: New line.

File: main/DatePicker/Calendar/Calendar.tsx [hunk 25]
```diff
@@ -0,0 +1,179 @@
...
+  const onViewableItemsChanged = React.useCallback(
+    ({ viewableItems, changed }) => {
+      if (viewableItems.length === 0) return; // 없는 경우 있음. 그냥 무시
+      const { index, isViewable, item, key } = viewableItems[0];
+      props.onChangeMonth && props.onChangeMonth(item);
+      setCurMonthFirst(item);
+    },
+    [],
+  );
```
Comment (JeffGuKang) [inline main/DatePicker/Calendar/Calendar.tsx]: New line.

File: main/DatePicker/Calendar/Calendar.tsx [hunk 26]
```diff
@@ -0,0 +1,179 @@
...
+    (monthFirst): React.ReactElement => {
+      console.log('>> monthFirst = ', monthFirst);
+      if (props.yearMonthComponent) {
+        return props.yearMonthComponent(monthFirst);
+      } else {
+        return <Text>{`${convertDateString(monthFirst)}`}</Text>;
+      }
+    },
+    [],
+  );
```
Comment (JeffGuKang) [inline main/DatePicker/Calendar/Calendar.tsx:115]: Empty line after paragraph code.

File: main/DatePicker/Calendar/CalendarDate.tsx [hunk 27]
```diff
@@ -0,0 +1,68 @@
...
+}
+
+function CalendarDate<T>(props: Props<T>): React.ReactElement {
+  const dateColor = props.date.getDay() === 0 ? 'red' : 'black';
+  const dateOpacity = props.isCurMonth ? 1 : 0.2;
+
+  const todayStyle = props.isToday
+    ? { borderRadius: 16, backgroundColor: '#efefef' }
+    : { borderRadius: 0, backgroundColor: 'white' };
+  return (
```
Comment (JeffGuKang) [inline main/DatePicker/Calendar/CalendarDate.tsx:41]: Empty line before return.

File: main/DatePicker/Calendar/CalendarDate.tsx [hunk 28]
```diff
@@ -0,0 +1,68 @@
...
+    <TouchableOpacity
+      onPress={(): void => {
+        props.onPress && props.onPress(props.date);
```
Comment (JeffGuKang) [inline main/DatePicker/Calendar/CalendarDate.tsx]: It is same with `props.onPress?.(props.date);` 👍

File: main/DatePicker/Calendar/CalendarMonth.tsx [hunk 29]
```diff
@@ -0,0 +1,110 @@
...
+  return dateString;
+};
+
+interface Props<T> {
+  monthDate: Date; // 렌더링 할 월의 날짜 (매월 1일)
+  style?: ViewStyle; // 스타일
+  dayComponent: ({
+    date,
+    dailyData,
+    isCurMonth,
```
Comment (JeffGuKang) [inline main/DatePicker/Calendar/CalendarMonth.tsx]: `isCurrentMonth` is more readable.

File: main/DatePicker/Calendar/CalendarDaysRow.tsx [hunk 30]
```diff
@@ -0,0 +1,58 @@
...
+
+const DaysRow = styled.View`
+  width: 100%;
+  height: 30px;
+  flex-direction: row;
+  align-items: center;
+  justify-content: space-between;
+  border-bottom-width: 1px;
+  border-bottom-color: #efefef;
+`;
```
Comment (JeffGuKang) [inline main/DatePicker/Calendar/CalendarDaysRow.tsx]: Need a new line.

File: main/DatePicker/Calendar/CalendarDaysRow.tsx [hunk 31]
```diff
@@ -0,0 +1,58 @@
...
+const DayColumn = styled.View`
+  width: 30px;
+  height: 100%;
+  align-items: center;
+  justify-content: center;
+`;
```
Comment (JeffGuKang) [inline main/DatePicker/Calendar/CalendarDaysRow.tsx]: Need a new line.

File: main/DatePicker/Calendar/CalendarDaysRow.tsx [hunk 32]
```diff
@@ -0,0 +1,58 @@
...
+const DayTitle = styled.Text`
+  font-size: 12px;
+  color: #565656;
+  line-height: 16px;
+`;
```
Comment (JeffGuKang) [inline main/DatePicker/Calendar/CalendarDaysRow.tsx]: New line.

File: main/DatePicker/Calendar/CalendarDaysRow.tsx [hunk 33]
```diff
@@ -0,0 +1,58 @@
...
+interface Props {
+  style?: ViewStyle;
+  calendarWidth: number;
+  weekdayFormat?: 'narrow' | 'short';
+}
+/**
+ * 캘린더 요일 row
+ */
+// const weekDays = ['일', '월', '화', '수', '목', '금', '토'];
+
```
Comment (JeffGuKang) [inline main/DatePicker/Calendar/CalendarDaysRow.tsx]: Use English and merge comments.

File: main/DatePicker/Calendar/CalendarDaysRow.tsx [hunk 34]
```diff
@@ -0,0 +1,58 @@
...
+function CalendarDaysRow(props: Props): React.ReactElement {
+  const weekDays: string[] = [];
+  for (let idx = 0; idx <= 6; idx++) {
+    const matchMonth = new Date(2020, 8, 6 + idx);
+    const weekDay = matchMonth.toLocaleString('default', {
+      weekday: props.weekdayFormat || 'narrow', // 'narrow',
+    });
+    weekDays.push(weekDay);
+  }
```
Comment (JeffGuKang) [inline main/DatePicker/Calendar/CalendarDaysRow.tsx]: New line.

File: main/DatePicker/Calendar/CalendarDaysRow.tsx [hunk 35]
```diff
@@ -0,0 +1,58 @@
...
+  weekdayFormat?: 'narrow' | 'short';
+}
+/**
+ * 캘린더 요일 row
+ */
+// const weekDays = ['일', '월', '화', '수', '목', '금', '토'];
+
+function CalendarDaysRow(props: Props): React.ReactElement {
+  const weekDays: string[] = [];
+  for (let idx = 0; idx <= 6; idx++) {
```
Comment (JeffGuKang) [inline main/DatePicker/Calendar/CalendarDaysRow.tsx]: New line.

File: main/DatePicker/Calendar/CalendarMonth.tsx [hunk 36]
```diff
@@ -0,0 +1,110 @@
...
+}
+/**
+ * Month component
+ */
+
+function CalendarMonth<T>(props: PropsWithChildren<Props<T>>): ReactElement {
+  // 금월 (시작월) 날짜
+  // const monthDate = props.monthDate ? moment(props.monthDate) : moment();
+  // 금월 달력의 시작 날짜 (= 금월 1일의 요일 수 만큼 뺌)
+  console.log('>>>> calendar width in CalendarMonth = ', props.calendarWidth);
```
Comment (JeffGuKang) [inline main/DatePicker/Calendar/CalendarMonth.tsx]: Do not leave a console log for users.

File: main/DatePicker/PickerCalendar.tsx [hunk 37]
```diff
@@ -0,0 +1,111 @@
...
+                    <View
+                      style={{
+                        height: 40,
+                        alignItems: 'center',
+                        justifyContent: 'center',
+                        backgroundColor: 'white',
+                      }}>
+                      <Text>{`${monthFirst.getFullYear()}년 ${
+                        monthFirst.getMonth() + 1
+                      }월`}</Text>
```
Comment (JeffGuKang) [inline main/DatePicker/PickerCalendar.tsx]: Express with only numbers will be better for users as `2020.5`

**Review by JeffGuKang — CHANGES_REQUESTED** [review]
It looks very useful feature. 
I tested manually on iOS simulator and found minor issues. 
Just fix some codes following coding convention.

- New line before/after new paragraphs.
- Do not leave `console.log`
- English is default language for dooboo-ui.

**Comment by JeffGuKang** [comment]
and `renderDay` is better than a `dayComponent` prop.

**Comment by JeffGuKang** [comment]
Refer the parameter names: https://github.com/wix/react-native-calendars

**Comment by JeffGuKang** [comment]
Test will be passed after increasing coverage more.

File: main/DatePicker/Calendar/Calendar.tsx (hunk 23 above)
Comment (bumjoo) [reply main/DatePicker/Calendar/Calendar.tsx]: I think it's not same.
This is to ensure that the component does not change even when all props are changed.

**Comment by JeffGuKang** [comment]
Good! Do not forget update snapshots.

**Review by JeffGuKang — APPROVED** [review]

File: .vscode/settings.json [hunk 38]
```diff
@@ -24,4 +24,5 @@
   "javascript.preferences.importModuleSpecifier": "relative",
   "typescript.preferences.importModuleSpecifier": "relative",
   "prettier.configPath": ".prettierrc.js",
+  "editor.formatOnSave": true,
```
Comment (JeffGuKang) [inline .vscode/settings.json:27]: This should be removed because it is a file that has nothing to do with this PR.

**Review by JeffGuKang — APPROVED** [review]
Great job! 
I will add a issue `Datepicker` feature to cover web environment.

File: .vscode/settings.json (hunk 38 above)
Comment (JeffGuKang) [reply .vscode/settings.json:27]: I will fix it so don't care about it.

**Pull request merged**